### PR TITLE
PTX-784 add command exec via nsenter to environments with ssh support disabled

### PR DESCRIPTION
**Pull request by thiguetta**

### CODE REVIEW

File: drivers/node/node.go [hunk 1]
```diff
@@ -27,6 +30,13 @@ const (
 	Directory FindType = "d"
 )
 
+const (
+	// PlatformGeneric identifies a generic node
+	PlatformGeneric PlatformType = "generic"
+	// PlatformIKS identifies a IKS node
+	PlatformIKS PlatformType = "iks"
```
Comment (harsh-px) [inline drivers/node/node.go]: nit: use "IKS"

File: drivers/node/ssh/ssh.go [hunk 2]
```diff
@@ -335,8 +293,67 @@ func (s *ssh) Systemctl(n node.Node, service string, options node.SystemctlOpts)
...
+		return s.doCmdSSH(n, options, cmd, ignoreErr)
+	}
+}
+
+func (s *ssh) doCmdIks(n node.Node, options node.ConnectionOpts, cmd string, ignoreErr bool) (string, error) {
+	cmdPreffix := []string{"--", "nsenter", "--mount=/hostproc/1/ns/mnt", "/bin/bash", "-c"}
+	cmdPreffix = append(cmdPreffix, fmt.Sprintf("\"%s\"", cmd))
+
+
+	pxPods, err := k8s.Instance().GetPodsByNode(n.Name, "kube-system")
```
Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: use constant from k8s package for "kube-system"

Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: It's not `pxPods` but rather debug pods.

File: tests/common.go [hunk 3]
```diff
@@ -84,6 +87,27 @@ func InitInstance() {
...
 	expect(err).NotTo(haveOccurred())
+
+	nodes := node.GetWorkerNodes()
+	expect(nodes).NotTo(beEmpty())
+
+	switch nodes[0].PlatformType{
+	case node.PlatformIKS:
+		specFactory, err := spec.NewFactory(fmt.Sprintf("%s/ssh", defaultSpecsRoot), Inst().S)
+		expect(err).NotTo(haveOccurred())
+		dsSpec, err := specFactory.Get("ssh")
```
Comment (harsh-px) [inline tests/common.go]: instead of ssh, we can call it debug since actually it's not doing ssh but rather nsenter.

File: tests/common.go [hunk 4]
```diff
@@ -84,6 +87,27 @@ func InitInstance() {
...
+		expect(err).NotTo(haveOccurred())
+		ds, err := k8s.Instance().CreateDaemonSet(dsSpec.SpecList[0].(*v1beta2.DaemonSet))
```
Comment (harsh-px) [inline tests/common.go]: You need to handle a case where the daemonset may already exist.

**Comment by harsh-px** [comment]
Can you fix the conflicts?

File: drivers/node/ssh/ssh.go [hunk 5]
```diff
@@ -335,20 +294,71 @@ func (s *ssh) Systemctl(n node.Node, service string, options node.SystemctlOpts)
...
+		return s.doCmdSSH(n, options, cmd, ignoreErr)
+	}
+}
+
+func (s *ssh) doCmdIks(n node.Node, options node.ConnectionOpts, cmd string, ignoreErr bool) (string, error) {
+	cmdPreffix := []string{"--", "nsenter", "--mount=/hostproc/1/ns/mnt", "/bin/bash", "-c"}
+	cmdPreffix = append(cmdPreffix, fmt.Sprintf("\"%s\"", cmd))
+
+
+	debugPods, err := k8s.Instance().GetPodsByNode(n.Name, iksDefaultNamespace)
```
Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: Sorry, rename `debugPods` to `allPodsForNode`.

File: drivers/node/ssh/ssh.go [hunk 6]
```diff
@@ -100,50 +178,40 @@ func (s *ssh) Init() error {
...
 
 func (s *ssh) TestConnection(n node.Node, options node.ConnectionOpts) error {
-	_, err := s.getAddrToConnect(n, options)
-	if err != nil {
+	var err error
+	var cmd string
+
+	switch n.PlatformType {
+	case node.PlatformIKS:
+		cmd = "cat /etc/hostname"
```
[the remark this answers is not in the document]
Comment (harsh-px) [reply drivers/node/ssh/ssh.go]: As as when possible, I want majority of torpedo to be platform agnostic.

File: drivers/node/ssh/ssh.go [hunk 7]
```diff
@@ -343,20 +361,73 @@ func (s *ssh) Systemctl(n node.Node, service string, options node.SystemctlOpts)
...
+	case node.PlatformIKS:
+		return s.doCmdIks(n, options, cmd, ignoreErr)
+	case node.PlatformGeneric:
+		fallthrough
+	default:
+		return s.doCmdSSH(n, options, cmd, ignoreErr)
+	}
+}
+
+func (s *ssh) doCmdIks(n node.Node, options node.ConnectionOpts, cmd string, ignoreErr bool) (string, error) {
```
Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: Rename to doCmdUsingPod to keep it generic

File: drivers/node/ssh/ssh.go [hunk 8]
```diff
@@ -55,6 +73,66 @@ func getKeyFile(keypath string) (ssh_pkg.Signer, error) {
 }
 
 func (s *ssh) Init() error {
+
+	nodes := node.GetWorkerNodes()
+	var err error
+
+	switch nodes[0].PlatformType {
```
Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: actually let's remove references to IKS itself.  Do this.

If torpedo is invoked with an empty ssh key and ssh password, then automatically install the ssh daemonset and configure the node driver to use that for all ssh operations.

End user should not needed to anything else.

**Comment by thiguetta** [comment]
test passed without ssh https://jenkins.portworx.co/job/Torpedo/view/Torpedo%202.1/job/tp-2-1-kops/105/console and with ssh https://jenkins.portworx.co/job/Torpedo/view/Torpedo%202.1/job/tp-2-1-kops/106/console

File: drivers/node/ssh/ssh.go [hunk 9]
```diff
@@ -57,7 +74,67 @@ func getKeyFile(keypath string) (ssh_pkg.Signer, error) {
...
+
+	return nil
+}
+
+func (s *ssh) initExecPod() error {
+	var ds *v1beta2.DaemonSet
+	var err error
+	k8sops := k8s.Instance()
+	if ds, err = k8sops.GetDaemonSet(execPodDaemonSetLabel, execPodDefaultNamespace); ds == nil {
+		logrus.Infof("error: %s", err)
```
Comment (harsh-px) [inline drivers/node/ssh/ssh.go]: looks like a debug log. remove if so.

**Pull request merged**